pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,251 @@
+	nameOfSchedule := w.goGfInterface.GetScheduleName(s)
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		_ = w.rateLimiter.Take()
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s.Name, catchupTime, err)
+			return err
+		} else {
+			w.snapshot.UpdateLastExecutionTime(nameOfSchedule, catchupTime)
+		}
+	}
+
+	return nil
+}
+
+func fire(ctx context.Context, executionManager mgInterfaces.ExecutionInterface, scheduledTime time.Time,
+	s models.SchedulableEntity) error {
+
+	literalsInputMap := map[string]*core.Literal{}
+	literalsInputMap[s.KickoffTimeInputArg] = &core.Literal{
+		Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: &core.Primitive{
+						Value: &core.Primitive_Datetime{
+							Datetime: timestamppb.New(scheduledTime),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	executionRequest := admin.ExecutionCreateRequest{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19],
+		Spec: &admin.ExecutionSpec{
+			LaunchPlan: &core.Identifier{
+				ResourceType: core.ResourceType_LAUNCH_PLAN,
+				Project:      s.Project,
+				Domain:       s.Domain,
+				Name:         s.Name,
+				Version:      s.Version,
+			},
+			Metadata: &admin.ExecutionMetadata{
+				Mode:        admin.ExecutionMetadata_SCHEDULED,
+				ScheduledAt: timestamppb.New(scheduledTime),
+			},
+			// No dynamic notifications are configured either.
+		},
+		// No additional inputs beyond the to-be-filled-out kickoff time arg are specified.
+		Inputs: &core.LiteralMap{
+			Literals: literalsInputMap,
+		},
+	}
+	if !*s.Active {
+		// no longer active
+		logger.Debugf(ctx, "schedule %v is no longer active", s.Name)
+		return nil
+	}
+	_, err := executionManager.CreateExecution(context.Background(), executionRequest, scheduledTime)
+	if err != nil {
+		logger.Error(ctx, "failed to create execution create request %v due to %v", executionRequest, err)
+		return err
+	}
+	return nil
+}
+
+func (w *workflowExecutor) Run() {
+	ctx := context.Background()
+	c, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		panic(fmt.Errorf("unable to run the workflow executor after reading the schedules due to %v", err))
+	}
+	// Start the go routine to write the snapshot map to the configured snapshot file
+	go w.CheckPointState(ctx)
+	catchUpTill := time.Now()
+	err = w.CatchUpAllSchedules(ctx, c.Entities, catchUpTill)
+	if err != nil {
+		panic(fmt.Errorf("unable to catch up on the schedules due to %v", err))
+	}
+
+	defer logger.Infof(ctx, "Exiting Workflow executor")
+	for true {
+		for _, s := range c.Entities {
+			funcRef := func() {
```

Comment:
lets pass in a context here

Reply:
and schedule time

Reply:
Done. Instead of schedule time i am passing the schedule and also the semantics of this func as explained in my previous comment is not to send execution for a single schedule time but over all the times between last execution time and current time when this func gets triggered.


That way we also are able to handle cases where some goroutine are running slow for some reason and hence we just use catch everytime